File: src/lib/python/package/lib_list.ts
import { Package, PackageState, type PackageData } from "./package";
import { pkgManager } from "./package_manager";
import { Webtty } from "../../websocket/websocket";

const MUST_TYPE = "must";
const OPTION_TYPE = "option";
const CACHE_TYPE = "cache";

interface PackageDict {
    [key: string]: { pack_type: string; index: number };
}

interface InstallTag {
    [key: string]: any;
}

interface PreState {
    [key: string]: PackageState;
}

export class PackageList {
    private state: PackageState = PackageState.installed;
    private queue: string[] = [];
    private tmpQueue: string[] = [];
    private currentInstalling: string | null = null;
    private packages: { [key: string]: Package[] } = {
        [MUST_TYPE]: [
            new Package(
                "xes-lib",
                "xes-lib是学而思专用的python库，实现了很多实用的功能，包括发送、路径查询、预处理、文件操作等功能...",
                PackageState.not_installed
            ),
            new Package(
                "qrcode",
                "二维码编码解码库。",
                PackageState.not_installed
            ),
        ],
        [OPTION_TYPE]: [
            new Package(
                "Pillow",
                "一个图像处理库，可以对图像进行旋转、缩放、裁剪、增色等处理。",
                PackageState.not_installed
            ),
            new Package(
                "numpy",
                "Python最常用的科学计算库，为Python提供了很多高级数学函数。",
                PackageState.not_installed
            ),
            new Package(
                "algorithms",
                "一个 Python 算法库，提供了常用的数据结构和算法。",
                PackageState.not_installed
            ),
        ],
        [CACHE_TYPE]: [],
    };
    private errDict: PackageDict = {};
    private nameDict: PackageDict = {};
    private staticList: { [key: string]: number } = {};
    private userList: Package[] = [];
    private descList: { [key: string]: string } = {};
    private installedErrQue: string[] = [];
    private hasNewErr: number = 0;
    private installTags: InstallTag = {};
    private preStates: PreState = {};
    private lockId: string | null = null;
    private lastReqTime: number | null = null;
    private webtty: Webtty | null = null;

    constructor() {
        for (const [key, list] of Object.entries(this.packages)) {
            list.forEach((item, index) => {
                this.nameDict[item.name] = { pack_type: key, index };
            });
        }

        this.startAutoDestroyState();
    }

    setWebtty(webtty: Webtty): void {
        this.webtty = webtty;
    }

    private startAutoDestroyState(): void {
        setInterval(() => {
            this.getState("");
        }, 1000);
    }

    async loadLocal(): Promise<void> {
        const localPkg = await pkgManager.getLocalList(true);
        this.userList = localPkg.user.map(
            (p: any) =>
                new Package(
                    p.name,
                    p.version || "",
                    PackageState.installed
                )
        );

        const userNameDict: { [key: string]: number } = {};
        for (const pack of this.userList) {
            userNameDict[pack.name] = 1;
            if (pack.name in this.errDict) {
                this.removeErrPack(pack.name);
            }
            if (pack.name in this.nameDict) {
                const optType = this.nameDict[pack.name].pack_type;
                if (optType === MUST_TYPE) {
                    const servPack = this.getPackageByName(pack.name);
                    if (
                        servPack.version &&
                        servPack.version !== "" &&
                        pack.version !== servPack.version
                    ) {
                        this.uninstallHandler(pack.name);
                    }
                    this.changeStateByName(pack.name, PackageState.installed);
                }
            }
        }

        for (const pack of localPkg.lib) {
            this.staticList[pack.name] = 1;
        }
    }

    async checkMust(): Promise<void> {
        for (const mustPack of this.packages[MUST_TYPE]) {
            if (mustPack.state === PackageState.not_installed) {
                console.log(`需要安装前置 ${mustPack.name}`);
                try {
                    await this.installHandler(mustPack.name, mustPack.version, mustPack.desc, null);
                } catch (e) {
                    console.error(`前置安装错误: ${e}`);
                }
            }
        }
    }

    updatePackData(jsonRes: any): void {
        try {
            const tmpNameDict: PackageDict = {};
            const tmpPackages: { [key: string]: Package[] } = {
                [MUST_TYPE]: [],
                [OPTION_TYPE]: [],
            };

            for (const [key, list] of Object.entries(jsonRes.data.libs)) {
                for (const dic of list as any[]) {
                    const pack = Package.fromDict(dic);
                    tmpNameDict[pack.name] = { pack_type: key, index: tmpPackages[key].length };
                    tmpPackages[key].push(pack);
                }
            }

            this.nameDict = tmpNameDict;
            this.packages[MUST_TYPE] = tmpPackages[MUST_TYPE];
            this.packages[OPTION_TYPE] = tmpPackages[OPTION_TYPE];

            this.packages[CACHE_TYPE].forEach((pack, index) => {
                this.nameDict[pack.name] = { pack_type: CACHE_TYPE, index };
            });
        } catch (e) {
            console.error("更新包时出现错误:", e);
        }
    }

    changeStateByName(name: string, state: PackageState): void {
        if (state === PackageState.installing) {
            this.currentInstalling = name;
        }
        const packType = this.nameDict[name].pack_type;
        const index = this.nameDict[name].index;
        this.packages[packType][index].changeState(state);
    }

    getPackageByName(name: string): Package {
        const packType = this.nameDict[name].pack_type;
        const index = this.nameDict[name].index;
        return this.packages[packType][index];
    }

    async getPackList(pageId: string): Promise<any> {
        if (!this.checkLock(pageId)) {
            return false;
        }

        const res: any = {};
        res.state = this.state.valueOf();

        const localPkg = await pkgManager.getLocalList(true);
        const optionList = localPkg.user;
        const localList = localPkg.lib;

        const needFilterKeys: { [key: string]: number } = {};
        for (const pack of localList) {
            needFilterKeys[pack.name] = 1;
        }

        const filteredList = optionList.filter(
            (pack: any) => !needFilterKeys[pack.name]
        );

        const lastOptionList: Package[] = [];
        const needShow: string[] = [];

        for (const pack of filteredList) {
            if (pack.name in this.errDict) {
                this.removeErrPack(pack.name);
            }
            if (pack.name in this.nameDict) {
                const optType = this.nameDict[pack.name].pack_type;
                if (optType === MUST_TYPE || optType === OPTION_TYPE) {
                    this.changeStateByName(pack.name, PackageState.installed);
                } else {
                    if (pack.name in this.descList) {
                        pack.desc = this.descList[pack.name];
                    } else {
                        needShow.push(pack.name);
                    }
                    lastOptionList.push(Package.fromDict(pack));
                }
            } else {
                if (pack.name in this.descList) {
                    pack.desc = this.descList[pack.name];
                } else {
                    needShow.push(pack.name);
                }
                lastOptionList.push(Package.fromDict(pack));
                this.nameDict[pack.name] = {
                    pack_type: CACHE_TYPE,
                    index: this.packages[CACHE_TYPE].length,
                };
                this.packages[CACHE_TYPE].push(Package.fromDict(pack));
            }
        }

        res[MUST_TYPE] = [...this.packages[MUST_TYPE]];
        res[OPTION_TYPE] = [...this.packages[OPTION_TYPE], ...lastOptionList];
        res.state = this.state.valueOf();

        if (needShow.length > 0) {
            this.checkPackDesc(needShow);
        }

        return this.filterRetLibs(res);
    }

    getErrList(): any[] {
        const res: any[] = [];
        for (const name in this.errDict) {
            const cache = this.nameDict[name];
            res.push(this.packages[cache.pack_type][cache.index]);
        }
        return res;
    }

    private filterRetLibs(originLibs: any): any {
        const targetLibs = { ...originLibs };
        const newMustArr = targetLibs[MUST_TYPE].filter(
            (lib: Package) => lib.name !== "xesrepair"
        );
        targetLibs[MUST_TYPE] = newMustArr;
        return targetLibs;
    }

    private async checkPackDesc(showList: string[]): Promise<void> {
        for (const name of showList) {
            if (!(name in this.descList)) {
                const info = await pkgManager.getModuleInfo(name);
                if (info.Summary) {
                    this.descList[name] = info.Summary;
                }
            }
        }
    }

    private checkLock(pageId: string): boolean {
        if (this.state === PackageState.installing) {
            if (this.lockId === null) {
                this.lockId = pageId;
                return true;
            }
            if (this.lockId === pageId) {
                return true;
            }
            if (pageId === null) {
                return true;
            }
            return false;
        } else {
            this.lockId = null;
            return true;
        }
    }

    unlock(pageId: string): void {
        if (this.lockId === pageId) {
            this.lockId = null;
        }
    }

    async getAllState(pageId: string): Promise<any> {
        if (!this.checkLock(pageId)) {
            return false;
        }

        const res: any = {};
        res.all_state = this.state.valueOf();
        res.err_count = Object.keys(this.errDict).length;
        let curCnt = 0;
        if (this.currentInstalling !== null) {
            curCnt = 1;
        }
        res.installing_count = curCnt + this.queue.length;
        return res;
    }

    getState(pre: string, isReq: boolean = false): any {
        if (isReq) {
            this.lastReqTime = Date.now();
        } else {
            const res = pkgManager.getProcess();
            if (res === null) {
                return {
                    all_state: this.state.valueOf(),
                    err_count: Object.keys(this.errDict).length,
                    installing_count: this.currentInstalling ? 1 + this.queue.length : this.queue.length,
                    desc: "",
                    has_new_err: false,
                    tag: null,
                };
            }

            if (res.state === "installed") {
                if (res.name in this.errDict) {
                    delete this.errDict[res.name];
                }
                this.changeStateByName(res.name, PackageState.installed);
                this.checkNext();
            } else if (res.state === "error") {
                this.hasNewErr = 1;
                this.errDict[res.name] = this.nameDict[res.name];
                this.changeStateByName(res.name, PackageState.err);
                this.checkNext();
            }

            return {
                ...res,
                all_state: this.state.valueOf(),
                err_count: Object.keys(this.errDict).length,
                installing_count: this.currentInstalling ? 1 + this.queue.length : this.queue.length,
            };
        }
    }

    private checkNext(): void {
        if (this.queue.length > 0) {
            const nextP = this.queue.shift()!;
            this.currentInstalling = nextP;
            this.changeStateByName(nextP, PackageState.installing);
            const pack = this.getPackageByName(nextP);
            pkgManager.handleInstall(pack);
        } else {
            this.currentInstalling = null;
            if (Object.keys(this.errDict).length > 0) {
                if (this.hasNewErr === 1) {
                    this.hasNewErr = 2;
                } else {
                    this.hasNewErr = 0;
                }
                this.state = PackageState.err;
            } else {
                this.state = PackageState.installed;
            }
        }
    }

    async installHandler(
        name: string,
        version: string = "",
        desc: string = "",
        pageId: string | null = null
    ): Promise<string> {
        this.installTags[name] = null;
        if (name !== "xesrepair") {
            this.state = PackageState.installing;
        } else {
            const canUse = this.checkLock(pageId || "");
            return canUse ? "true" : "false";
        }

        if (!(name in this.nameDict)) {
            this.preStates[name] = PackageState.not_installed;
            const cachePack = new Package(name, desc);
            this.packages[CACHE_TYPE].push(cachePack);
            this.nameDict[name] = {
                pack_type: CACHE_TYPE,
                index: this.packages[CACHE_TYPE].length - 1,
            };
        } else {
            const packType = this.nameDict[name].pack_type;
            const index = this.nameDict[name].index;
            const cachePack = this.packages[packType][index];
            this.preStates[name] = PackageState.not_installed;
            if (cachePack.state === PackageState.err) {
                this.preStates[name] = PackageState.err;
            }
        }

        if (this.queue.length > 0 || this.currentInstalling !== null) {
            this.queue.push(name);
            this.changeStateByName(name, PackageState.waiting);
            return PackageState.waiting.valueOf();
        }

        this.changeStateByName(name, PackageState.installing);
        const pack = this.getPackageByName(name);
        await pkgManager.handleInstall(pack);
        return PackageState.installing.valueOf();
    }

    async cancelInstallHandler(name: string): Promise<void> {
        const preState = this.preStates[name];
        this.changeStateByName(name, preState);

        if (name === this.currentInstalling) {
            pkgManager.cancelInstall();
            this.checkNext();
        } else {
            const idx = this.queue.indexOf(name);
            if (idx > -1) {
                this.queue.splice(idx, 1);
            }
        }
    }

    async uninstallHandler(name: string): Promise<void> {
        if (name in this.nameDict) {
            await pkgManager.handleUninstall(name);
            this.changeStateByName(name, PackageState.not_installed);
        }
    }

    private removeErrPack(name: string): void {
        if (name in this.errDict) {
            const cache = this.nameDict[name];
            this.packages[cache.pack_type][cache.index].state = PackageState.not_installed;
            delete this.errDict[name];
            if (this.queue.length === 0 && Object.keys(this.errDict).length === 0) {
                this.state = PackageState.installed;
            }
        }
    }

    async searchHandler(name: string, flag: boolean): Promise<any> {
        let resList: Package[] = [];

        if (flag && name in this.nameDict) {
            resList = [this.getPackageByName(name)];
        } else {
            const searchResult = await pkgManager.handleSearch(name);
            resList = searchResult.map((p: any) =>
                Package.fromDict({ name: p.name, desc: "", ...p })
            );
        }

        const res: any = { [MUST_TYPE]: [], [OPTION_TYPE]: [] };

        for (const pack of resList) {
            if (pack.name in this.staticList) {
                pack.state = PackageState.builtin;
                res[OPTION_TYPE].push(pack);
            } else if (pack.name in this.nameDict) {
                const cachePack = this.getPackageByName(pack.name);
                if (this.nameDict[pack.name].pack_type === MUST_TYPE) {
                    res[MUST_TYPE].push(cachePack);
                } else {
                    res[OPTION_TYPE].push(cachePack);
                }
            } else {
                res[OPTION_TYPE].push(pack);
            }
            if (flag) {
                this.packages[CACHE_TYPE].push(pack);
                this.nameDict[pack.name] = {
                    pack_type: CACHE_TYPE,
                    index: this.packages[CACHE_TYPE].length - 1,
                };
            }
        }

        return res;
    }
}

export const packList = new PackageList();

export async function loadOriginPack(): Promise<void> {
    try {
        const res = await fetch("http://code.xueersi.com/api/python/libs");
        if (res.ok) {
            const resObj = await res.json();
            packList.updatePackData(resObj);
            await packList.loadLocal();
            await packList.checkMust();
        } else {
            await packList.loadLocal();
        }
    } catch (e) {
        console.error("加载原始包出错：", e);
        await packList.loadLocal();
    }
}

File: lib_list.py
# uncompyle6 version 3.8.0
# Python bytecode 3.7.0 (3394)
# Decompiled from: Python 3.8.5 (tags/v3.8.5:580fbb0, Jul 20 2020, 15:57:54) [MSC v.1924 64 bit (AMD64)]
# Embedded file name: build\Lib\site-packages\thonny\lib_list.py
# Compiled at: 2022-12-29 15:08:18
# Size of source mod 2**32: 22634 bytes
from collections import deque
from enum import Enum
from thonny.manage_package import pkg_manager, empty_process_manager
from thonny.package import Package, State, init_pack_with_dict
from thonny.common import is_client_call, PIP_NAMES
from thonny.port_helper import check
import requests
from threading import Thread, Lock
from thonny.common import pre_path, set_lib_obj, get_lib_obj
from thonny.log import push_log, Type
import pickle, os, asyncio, queue, json, time
must_type = 'must'
option_type = 'option'
cache_type = 'cache'
pip_names = {}
lock = Lock()
res_obj = None

def filter_ret_libs(origin_libs):
    """
    :功能:过滤返回前端的库列表，如xesrepair属于必备库，但是不显示给用户
    """
    import copy
    target_libs = copy.deepcopy(origin_libs)
    new_must_arr = []
    for lib in target_libs[must_type]:
        if lib.name != 'xesrepair':
            new_must_arr.append(lib)

    target_libs[must_type] = new_must_arr
    return target_libs


class PackageList:

    def __init__(self):
        self.state = State.installed
        self._queue = deque()
        self._tmp_queue = deque()
        self._cur_installing = None
        self.packages = {}
        self.packages[must_type] = [
         Package(name='xes-lib',
           desc='xes-lib是学而思编程发布的python库，可以实现很多炫酷功能，如短信发送、路径查询、天气预报、语音翻译等功能，持续更新中.....'), Package(name='qrcode', desc='一个二维码生成工具库。专属二维码了解一下。')]
        self.packages[option_type] = [Package(name='Pillow', desc='一个基础图像处理库，可以对图像切片、旋转、滤镜、输出文字、调色。'),
         Package(name='numpy',
           desc='我们怎么能缺少这么重要的库？它为Python提供了很多高级的数学方法。'), Package(name='algorithms', desc='一个 Python 算法模块，提供了常用的数据结构及算法。')]
        self.packages[cache_type] = []
        self._err_dic = {}
        self._name_dic = {}
        self._static_list = {}
        self._user_list = None
        self._desc_list = {}
        self._installed_err_que = deque()
        self._has_new_err = 0
        self.install_tags = {}
        self.install_pre_states = {}
        self._lock_id = None
        self._last_req_time = None
        for key, l in self.packages.items():
            for index, item in enumerate(l):
                self._name_dic[item.name] = {'pack_type':key, 
                 'index':index}

        new_thread = Thread(target=(self.auto_destory_state), daemon=True)
        new_thread.start()

    def load_local(self):
        global res_obj
        local_pkg = pkg_manager.get_local_list(True)
        self._user_list = local_pkg['user']
        user_name_dic = {}
        for pack in self._user_list:
            user_name_dic[pack.name] = 1
            if pack.name in self._err_dic:
                self.remove_err_pack(pack.name)
            if pack.name in self._name_dic:
                opt_type = self._name_dic[pack.name]['pack_type']
            if opt_type == must_type:
                serv_pack = self.get_package_by_name(pack.name)
                if serv_pack.version is not None:
                    if serv_pack.version != '':
                        if pack.version != serv_pack.version and not is_client_call():
                            if 'open' in res_obj['data']:
                                if res_obj['data']['open'] == 1:
                                    self.uninstall_handler(pack.name)
                self.change_state_by_name(pack.name, State.installed)

        for pack in local_pkg['lib']:
            self._static_list[pack.name] = 1

    def check_must(self):
        for must_pack in self.packages[must_type]:
            if must_pack.state == State.not_installed.value:
                push_log(Type.LOG, 0, 'need install {}'.format(must_pack.name))
                try:
                    self.install_handler(name=(must_pack.name),
                      version=(must_pack.version),
                      desc=(must_pack.desc),
                      page_id=None)
                except Exception as e:
                    try:
                        push_log(Type.LOG, 0, 'must install err {}'.format(e))
                    finally:
                        e = None
                        del e

    def update_pack_data(self, json_res):
        global pip_names
        try:
            tmp_name_dic = {}
            tmp_packages = {must_type: [], option_type: []}
            pip_names = json_res['data']['pip_names']
            for key, l in json_res['data']['libs'].items():
                for index, dic in enumerate(l):
                    pack = init_pack_with_dict(dic)
                    tmp_name_dic[pack.name] = {'pack_type':key, 
                     'index':index}
                    tmp_packages[key].append(pack)

            self._name_dic = tmp_name_dic
            self.packages[must_type] = tmp_packages[must_type]
            self.packages[option_type] = tmp_packages[option_type]
            for index, pack in enumerate(self.packages[cache_type]):
                self._name_dic[pack.name] = {'pack_type':cache_type, 
                 'index':index}

        except Exception as e:
            try:
                push_log((Type.LOG), status=0, msg=e)
            finally:
                e = None
                del e

    def load_desc(self):
        desc_path = os.path.join(pre_path, 'desc.txt')
        if not os.path.exists(desc_path):
            f = open(desc_path, 'wb')
            pickle.dump({}, f, -1)
            f.close()
        else:
            f = open(desc_path, 'rb')
            self._desc_list = pickle.load(f)
            f.close()
            need_show = []
            if self._user_list is None:
                local_pkg = pkg_manager.get_local_list()['user']
            else:
                local_pkg = self._user_list
        for pack in local_pkg:
            if pack.name not in self._desc_list:
                need_show.append(pack.name)

        if len(need_show) > 0:
            self.check_pack_desc(need_show)

    def get_pack_show(self, name):
        show_dict = pkg_manager.get_module_info(name)
        if 'Summary' in show_dict:
            summary = show_dict['Summary']
            self._desc_list[name] = summary

    def check_pack_desc(self, show_list):
        for name in show_list:
            if name not in self._desc_list:
                self.get_pack_show(name)

        with lock:
            desc_path = os.path.join(pre_path, 'desc.txt')
            f = open(desc_path, 'wb')
            pickle.dump(self._desc_list, f, -1)
            f.close()

    def change_state_by_name(self, name, state):
        if state == State.installing:
            self._cur_installing = name
        pack_type = self._name_dic[name]['pack_type']
        index = self._name_dic[name]['index']
        self.packages[pack_type][index].change_state(state)

    def get_package_by_name(self, name):
        pack_type = self._name_dic[name]['pack_type']
        index = self._name_dic[name]['index']
        return self.packages[pack_type][index]

    def get_state_by_name(self, name):
        return self.get_state_by_name(name).state

    def get_pack_list(self, page_id):
        can_use = self.check_lock(page_id)
        if not can_use:
            return False
        res = {}
        res['state'] = self.state.value
        local_pkg = pkg_manager.get_local_list(True)
        option_list = local_pkg['user']
        local_list = local_pkg['lib']
        need_show = []
        last_option_list = []
        filtered_list = []
        need_filter_keys = {}
        for pack in local_list:
            need_filter_keys[pack.name] = 1

        for pack in option_list:
            if pack.name not in need_filter_keys:
                filtered_list.append(pack)

        for pack in filtered_list:
            if pack.name in self._err_dic:
                self.remove_err_pack(pack.name)
            if pack.name in self._name_dic:
                opt_type = self._name_dic[pack.name]['pack_type']
                if opt_type == must_type or opt_type == option_type:
                    self.change_state_by_name(pack.name, State.installed)
                else:
                    if pack.name in self._desc_list:
                        pack.desc = self._desc_list[pack.name]
                    else:
                        need_show.append(pack.name)
                    last_option_list.append(pack)
            else:
                if pack.name in self._desc_list:
                    pack.desc = self._desc_list[pack.name]
                else:
                    need_show.append(pack.name)
                last_option_list.append(pack)
                self._name_dic[pack.name] = {'pack_type':'cache', 
                 'index':len(self.packages[cache_type])}
                self.packages[cache_type].append(pack)

        res[must_type] = self.packages[must_type].copy()
        res[option_type] = self.packages[option_type].copy()
        res[option_type].extend(last_option_list)
        res['state'] = self.state.value
        if len(need_show) > 0:
            update_thread = Thread(target=(self.check_pack_desc),
              args=(need_show,),
              daemon=True)
            update_thread.start()
        res = filter_ret_libs(res)
        return res

    def get_err_list(self):
        res = []
        for name in self._err_dic:
            cache = self._name_dic[name]
            res.append(self.packages[cache['pack_type']][cache['index']])

        return res

    def auto_destory_state(self):
        time.sleep(1)
        while True:
            now = time.time()
            if not self.state == State.installing or self._last_req_time is None or now - self._last_req_time > 1:
                self.get_state('')
                time.sleep(0.3)
            else:
                time.sleep(1)

    def get_state(self, pre, is_req=False):
        if is_req:
            self._last_req_time = time.time()
        else:
            res = pkg_manager.get_process()
            if res is None:
                res = empty_process_manager
                res.all_state = self.state.value
                res.err_count = len(self._err_dic)
                cur_cnt = 0
                if self._cur_installing is not None:
                    cur_cnt = 1
                else:
                    res.installing_count = cur_cnt + len(self._queue)
                    res.desc = ''
                    res.has_new_err = False
                    res.tag = None
                    return res
                    res.err_count = len(self._err_dic)
                    res.installing_count = 1 + len(self._queue)
                    res.all_state = State.installing.value
                    name = res.name
                    res.desc = self.get_package_by_name(name).desc
                    if res.state == State.installed.value:
                        if name in self._err_dic:
                            self._err_dic.pop(name)
                        self.change_state_by_name(name, State.installed)
                        self.check_next()
                        res.all_state = self.state.value
                        self._installed_err_que.append(res)
                    else:
                        if res.state == State.err.value:
                            self._has_new_err = 1
                            self._err_dic[name] = self._name_dic[name]
                            self.change_state_by_name(name, State.err)
                            self.check_next()
                            res.all_state = self.state.value
                            self._installed_err_que.append(res)
                        else:
                            res.all_state = self.state.value
                if self._has_new_err == 2:
                    res.has_new_err = True
                    self._has_new_err = 0
            else:
                res.has_new_err = False
        res.tag = self.install_tags[name]
        return res

    def check_lock(self, page_id):
        if self.state == State.installing:
            if self._lock_id is None:
                self._lock_id = page_id
                return True
            if self._lock_id == page_id:
                return True
            if page_id == None:
                return True
            return False
        else:
            self._lock_id = None
            return True

    def unlock(self, page_id):
        if self._lock_id == page_id:
            self._lock_id = None

    def get_all_state(self, page_id):
        can_use = self.check_lock(page_id)
        if not can_use:
            return False
        res = {}
        res['all_state'] = self.state.value
        res['err_count'] = len(self._err_dic)
        cur_cnt = 0
        if self._cur_installing is not None:
            cur_cnt = 1
        res['installing_count'] = cur_cnt + len(self._queue)
        return res

    def check_next(self):
        if self._queue:
            nextp = self._queue.popleft()
            self._cur_installing = nextp
            self.change_state_by_name(nextp, State.installing)
            pack = self.get_package_by_name(nextp)
            pkg_manager.handle_install(pack)
        else:
            self._cur_installing = None
            if len(self._err_dic):
                if self._has_new_err == 1:
                    self._has_new_err = 2
                else:
                    self._has_new_err = 0
                self.state = State.err
            else:
                self.state = State.installed

    def install_handler(self, name, page_id, version='', desc='', first=False, tag=None):
        self.install_tags[name] = tag
        if name != 'xesrepair':
            self.state = State.installing
        else:
            can_use = self.check_lock(page_id)
            return can_use or False
        if name not in self._name_dic:
            self.install_pre_states[name] = State.not_installed
            cache_pack = Package(name=name, desc=desc)
            self.packages[cache_type].append(cache_pack)
            self._name_dic[name] = {'pack_type':'cache',  'index':len(self.packages[cache_type]) - 1}
        else:
            pack_type = self._name_dic[name]['pack_type']
            index = self._name_dic[name]['index']
            cache_pack = self.packages[pack_type][index]
            self.install_pre_states[name] = State.not_installed
            if cache_pack.state == State.err.value:
                self.install_pre_states[name] = State.err
        if self._queue or self._cur_installing is not None:
            if first:
                self._queue.appendleft(name)
            else:
                self._queue.append(name)
            self.change_state_by_name(name, State.waiting)
            return State.waiting.value
        self.change_state_by_name(name, State.installing)
        pkg_manager.handle_install(cache_pack)
        return State.installing.value

    def cancel_install_handler(self, name):
        pre_state = self.install_pre_states[name]
        self.change_state_by_name(name, self.install_pre_states[name])
        if name == self._cur_installing:
            pkg_manager.handle_cancel_install()
            self.check_next()
        else:
            while self._queue:
                top = self._queue.popleft()
                if top == name:
                    break
                else:
                    self._tmp_queue.append(top)

            while self._tmp_queue:
                top = self._tmp_queue.pop()
                self._queue.appendleft(top)

        return pre_state.value

    def _filter_search_list(self, old_search_list, search_name):
        """
        对搜索列表进行过滤，根据库名映射配置优先展示必备库
        :params old_search_list 已有的返回
        :params search_name 搜索库名
        """
        for package in old_search_list:
            if package.name == 'xes':
                old_search_list.remove(package)
                break

        if search_name not in pip_names:
            return old_search_list
        for package in old_search_list:
            if package.name == pip_names[search_name]:
                return old_search_list

        for package in self._user_list:
            if package.name == pip_names[search_name]:
                old_search_list.insert(0, package)
                return old_search_list

        return old_search_list

    def search_handler(self, name, flag):
        if flag and name in self._name_dic:
            res_list = [
             self.get_package_by_name(name)]
        else:
            res_list = pkg_manager.handle_search(name, flag)
            if flag == False:
                res_list = self._filter_search_list(res_list, name)
        res = {must_type: [], option_type: []}
        for pack in res_list:
            if pack.name in self._static_list:
                pack.state = State.builtin.value
                res[option_type].append(pack)
            elif pack.name in self._name_dic:
                cache_pack = self.get_package_by_name(pack.name)
                if self._name_dic[pack.name]['pack_type'] == must_type:
                    res[must_type].append(cache_pack)
                else:
                    res[option_type].append(cache_pack)
            else:
                res[option_type].append(pack)
            if flag:
                self.packages[cache_type].append(pack)
                self._name_dic[pack.name] = {'pack_type':'cache', 
                 'index':len(self.packages[cache_type]) - 1}

        return res

    def uninstall_handler(self, name):
        if name in self._name_dic:
            pkg_manager.handle_uninstall(name)
            self.change_state_by_name(name, State.not_installed)

    def remove_err_pack(self, name):
        if name in self._err_dic:
            cache = self._name_dic[name]
            self.packages[cache['pack_type']][cache['index']].state = State.not_installed.value
            self._err_dic.pop(name)
            if not self._queue:
                if len(self._err_dic) == 0:
                    self.state = State.installed


pack_list = PackageList()

def load_origin_pack():
    new_thread = Thread(target=req_origin_pack, daemon=True)
    new_thread.start()
    load_thread = Thread(target=(pack_list.load_desc), daemon=True)
    load_thread.start()


def req_origin_pack():
    global res_obj
    url = 'http://code.xueersi.com/api/python/libs'
    res = requests.get(url, timeout=10)
    if res.status_code == 200:
        res_obj = res.json()
        pkg_manager.set_lib_obj(res_obj)
        set_lib_obj(res_obj)
        pack_list.update_pack_data(res_obj)
        pack_list.load_local()
        if not is_client_call():
            if not 'open' in res_obj['data'] or res_obj['data']['open'] == 1:
                pack_list.check_must()
    else:
        pack_list.load_local()
# okay decompiling lib_list.pyc
